src/GraphQL/Types/Composite/InputObjectGraphType.cs

```diff
@@ -60,7 +61,41 @@ public virtual object ParseDictionary(IDictionary<string, object?> value)
         }
 
         /// <inheritdoc/>
-        public virtual bool IsValidDefault(object value) => value is TSourceType;
+        public virtual bool IsValidDefault(object value)
+        {
+            if (value is not TSourceType)
+                return false;
+
+            foreach (var field in Fields)
```

Comment:
```suggestion
            if (GetType() != typeof(InputObjectGraphType<TSourceType>)
                return true;

            foreach (var field in Fields)
```

How about this to account for if people derive from this class in order to change `ParseDictionary`? Otherwise every derived instance probably won't work correctly.

Reply:
I would not.

ParseDictionary:

> Overriding this method allows for customizing the deserialization process of input objects, much like a field resolver does for output objects.

`IsValidDefault` works for some already constructed runtime value. I want to say that these two methods are not connected with each other and are not required to correlate. Such a default implementation is quite simple and makes sense. In the end, both of them are virtual, so that if there are problems, they can always be overridden.

Reply:
> Otherwise every derived instance probably won't work correctly.

?

Reply:
Any time that ParseDictionary is used, it is for the purpose of mutating the inputs from the AST or variables into the target object.  IsValidDefault checks the value against the AST.  It would be required to "unmutate" the values in order to check it against the AST.  So pretty much by definition, IsDefaultValue will be incorrect if ParseDictionary is used.  Further, IsDefaultValue is not generally necessary, and it should not be necessary to implement a complicated IsDefaultValue method just to use ParseDictionary.

If anything, we should be implementing a ToDictionary method, which IsDefaultValue can use to check the dictionary against the AST.  This would be much clearer of its intent and how to properly implement it.  But it would only ever be used for validating a default value, and seems rather pointless.

Reply:
I'm not sure if you've used ParseDictionary much, or read up on how other people are using it.  But I believe it's often when the shape of the input object does not match the shape of TObjectType.  Of course it could be used just for validating field values or similar.

Reply:
>  we should be implementing a ToDictionary method, which IsDefaultValue can use to check the dictionary against the AST. 

How `IsDefaultValue` can use it if it's just a dictionary?